src/dispatch/dispatch_virtual.h

```diff
@@ -0,0 +1,194 @@
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact darma@sandia.gov
+//
+// *****************************************************************************
+//@HEADER
+*/
+
+#if !defined(CHECKPOINT_DISPATCH_VIRTUAL_H)
+#define CHECKPOINT_DISPATCH_VIRTUAL_H
+
+#include "dispatch.h"
+
+#include <vector>
+#include <tuple>
+#include <functional>
+
+namespace serdes {
+  using AutoHandlerType = int;
+
+  struct SERIALIZE_CONSTRUCT_TAG {};
+
+  template <typename T, typename SerializerT>
+  using RegistryType = std::vector<
+    std::tuple<
+      int,
+      std::function<T*(void)>
+      >
+    >;
+
+  template <typename T, typename SerializerT>
+  inline RegistryType<T, SerializerT>& getRegistry();
+
+  template <typename ObjT, typename SerializerT>
+  struct Registrar {
+    Registrar();
+    AutoHandlerType index;
+  };
+
+  template <typename ObjT, typename SerializerT>
+  struct Type {
+    static AutoHandlerType const idx;
+  };
+
+  inline AutoHandlerType getObjIdx(AutoHandlerType han);
+
+  template <typename ObjT, typename SerializerT>
+  inline AutoHandlerType makeObjIdx();
+
+
+  template <typename T, typename SerializerT>
+  inline RegistryType<T,SerializerT>& getRegistry() {
+    static RegistryType<T,SerializerT> reg;
+    return reg;
+  }
+
+  template <typename ObjT, typename SerializerT>
+  Registrar<ObjT, SerializerT>::Registrar() {
+    using BaseType = typename ObjT::SerDerBaseType;
+
+    auto& reg = getRegistry<BaseType, SerializerT>();
+    index = reg.size();
+
+    debug_serdes("registrar: %ld, %s %s\n", reg.size(), typeid(ObjT).name(), typeid(SerializerT).name());
+
+    reg.emplace_back(
+		     std::make_tuple(
+				     index,
+				     []() -> BaseType* { return new ObjT(SERIALIZE_CONSTRUCT_TAG{}); }
+				     )
+		     );
+  }
+
+  template <typename ObjT, typename SerializerT>
+  AutoHandlerType const Type<ObjT, SerializerT>::idx = Registrar<ObjT, SerializerT>().index;
+
+  template <typename T, typename SerializerT>
+  inline auto getObjIdx(AutoHandlerType han) {
+    debug_serdes("getObjIdx: han=%d, size=%ld\n", han, getRegistry<T, SerializerT>().size());
+    return getRegistry<T, SerializerT>().at(han);
+  }
+
+  template <typename ObjT, typename SerializerT>
+  inline AutoHandlerType makeObjIdx() {
+    return Type<ObjT, SerializerT>::idx;
+  }
+
+  template <typename DerivedT, typename BaseT>
+  struct SerializableDerived : BaseT {
+    using SerDerBaseType = BaseT;
+
+    template <typename... Args>
+    SerializableDerived(Args&&... args)
+      : BaseT(std::forward<Args>(args)...)
+    { }
+
+    SerializableDerived() {
+      // Must pretend to call this as so so it gets instantiated
+      if (false) {
+	neverRun();
+      }
+    }
+
+    template <typename SerializerT>
+    void serialize(SerializerT& s) {
+      debug_serdes("SerializableDerived:: serialize\n");
```

Comment:
This currently generates a warning `ISO C++11 requires at least one argument for the "..." in a variadic macro`